### PR TITLE
Update to v1.4.0

### PR DESCRIPTION
**Pull request by mmoreno-cart**
**Features**

- Covariate CSV input format: it allows now several covariates for a single file in one row instead of the old (but still supported) one covariate-one row. Optimises the computation as all the covariates in the row are obtained at the same zonal stat function call. Summary Statistics and Variable Name columns must have the same number of elements, delimited by semicolons.
- CSV delimiters. The tool now autodetects the delimiters in the input CSV files. Instead of just comma (,), it also supports now semicolon (;), tab and pipe (|). When semicolon is used at CSV file level, the delimiter for the covariates in the previous point must be the comma.
- Application of pole of inaccessibility instead of centroids when possible. It uses now pole of inaccessibility for nearest distance calculation (as seen more logical when working with irregular polygons like admin areas instead of just buffers).
- Added progress bar for steps 2 (displace) and 3 (compute covariates).
- Added button to open displaced anonymised centroids CSV file in step 2.

**Bugfixes**

- Pole of inaccessibility bug in CentroidsLoader (when loading multipoint clusters through CSV)
- Solved errors and warnings shown in Logs (when calling twice the centroid loader or the displacer)
- Replaced deprecated QgsField constructor using QVariant by QMetaType.Type

**Notes**
README.md has been combined with updates done during the new version development.
metadata.txt has been updated also in the zip file (otherwise QGIS users won't have it)

### CODE REVIEW

File: plugin/metadata.txt [hunk 1]
```diff
@@ -6,7 +6,7 @@
 name=MICS GEOCODE PLUGIN
 qgisMinimumVersion=3.0
```
Comment (jpazvd) [inline plugin/metadata.txt:7]: **🔴 Blocker — this version floor is incorrect and will crash older QGIS.**

This PR introduces APIs newer than 3.0:
- `QgsField(name, QMetaType.Type.X)` (used across `CentroidsLoader` / `CentroidsDisplacer` / `CentroidBuffersLayerWriter` / `CovariatesProcesser`) requires **QGIS 3.38+** — the `QVariant` constructor was only *deprecated* there, not replaced earlier.
- `Utils.writeLayerIfExists` now calls `writeAsVectorFormatV3`, which requires **QGIS 3.20+**.

On QGIS 3.0–3.37 the plugin raises `AttributeError` at the first layer creation. Because the same wrong floor is inside the hand-edited `metadata.txt` in `releases/micsgeocodeplugin-1.4.0.zip`, an older install passes the version gate and then fails at step 1.

**Fix:** set `qgisMinimumVersion=3.38`, note the floor change in `release-notes.md` (and the "QGIS 3.0+" line in `README.md`), and regenerate the release zip from `plugin/` via `WIN_zip.bat` so the shipped manifest matches.

File: plugin/micsgeocode/CentroidsLoader.py [hunk 2]
```diff
@@ -190,31 +192,10 @@ def __loadCentroidsFromSHP(self) -> typing.NoReturn:
...
-
-                if cluster_polygon.geometry().poleOfInaccessibility(pole_inaccessibility_precision)[0].isNull():
-                    cluster_centroid_ft.setGeometry(cluster_polygon.geometry().centroid())
-                else:
-                    cluster_centroid_ft.setGeometry(cluster_polygon.geometry().poleOfInaccessibility(pole_inaccessibility_precision)[0])
-
-                # cluster_centroid_ft.setGeometry(cluster_polygon.geometry().poleOfInaccessibility(100)[0])
+                #print(f"[CentroidsLoader] Computing centroid for cluster {cluster_polygon[self.cluster_no_field]}")
+
+                cluster_centroid_ft.setGeometry(Utils.getPoleOfInaccessibilityOrCentroid(cluster_polygon, self.pole_inaccessibility_precision))
```
Comment (jpazvd) [inline plugin/micsgeocode/CentroidsLoader.py:197]: **🔴 Blocker — precision is passed into the `fallback_feature` parameter.**

The helper is `getPoleOfInaccessibilityOrCentroid(polygon_feature, fallback_feature=None, precision=0.00001)`, so here `self.pole_inaccessibility_precision` binds to **`fallback_feature`**, not `precision`. Two consequences:

1. On the null-pole fallback branch the helper runs `fallback_feature.geometry()` on a `float` → `AttributeError`, crashing SHP centroid-loading for exactly the degenerate polygons the pole method is meant to handle.
2. The configurable `pole_inaccessibility_precision` is silently ignored on this path — it "works" only because the default happens to equal `0.00001`.

The multipoint call site (`__computeCentroidGeometry`) already passes these by keyword.

**Fix:**
```python
cluster_centroid_ft.setGeometry(
    Utils.getPoleOfInaccessibilityOrCentroid(
        cluster_polygon, precision=self.pole_inaccessibility_precision))
```
and add a test exercising the null-pole fallback branch. (Separately, `getPoleOfInaccessibilityOrCentroid` is annotated `-> typing.NoReturn` but returns a `QgsGeometry`.)

File: plugin/micsgeocode/CovariatesProcesser.py [hunk 3]
```diff
@@ -115,73 +112,102 @@ def computeCovariates(self) -> typing.NoReturn:
...
-                rowIndex = rowIndex + 1
+                    variety_index = sum_stats.index('variety')
+                    sum_stats.remove('variety')
+                    column_names.pop(variety_index)
+                    if len(sum_stats) == 0:
+                        progress_bar.update(int((i) / total_rows * 100))
+                        continue
+                
+                if len(sum_stats) != len(column_names):
+                    raise ValueError(f"In row {rowIndex+1} of the input CSV file, the number of summary statistics ({len(sum_stats)}) does not match the number of output columns ({len(column_names)}). Please correct the input file.")
```
Comment (jpazvd) [inline plugin/micsgeocode/CovariatesProcesser.py:192]: **🟠 Wrong row number, ordering, and inconsistent error policy.**

- `rowIndex` here is stale — it was left at `len(inputs)+1` after the parse loop — so every mismatch reports the **last** row regardless of which row actually failed. Use the per-row `row_index` you already extract; the "Unknown file format … row {row_index+1}" message a few branches below does this correctly.
- This `raise` aborts the **entire** batch, whereas unknown-format / unsupported-Shapefile-stat / all-`variety` rows soft-skip with `output_warning` + `continue`. Pick one contract.
- The `variety` handling above (`sum_stats.index('variety')` → `column_names.pop(variety_index)`) runs **before** this length check, so a genuinely mismatched row can `IndexError` there first. Move this `len(sum_stats) != len(column_names)` check above the variety block.

**Fix:** `f"In row {row_index+1} …"`, reorder the validation before the variety pop, and align the mismatch path with the soft-skip policy (or make every invalid-input case fatal — but consistently).

File: plugin/micsgeocode/Utils.py [hunk 4]
```diff
@@ -185,7 +194,21 @@ def getval(ft: QgsFeature, field: QgsField) -> str:
...
+# def detect_csv_delimiter(f, sample_lines=3):
+    """Auto-detect CSV delimiter by analyzing first few lines"""
+
+    with open(file_path, 'r', encoding='utf-8-sig') as f:
+        sample = ''.join([f.readline() for _ in range(sample_lines)])
+    #sample = ''.join([f.readline() for _ in range(sample_lines)])
 
+    sniffer = csv.Sniffer()
+    try:
+        dialect = sniffer.sniff(sample, delimiters=',;\t\|')
```
Comment (jpazvd) [inline plugin/micsgeocode/Utils.py:207]: **🟠 Invalid escape makes backslash a candidate delimiter.**

`',;\t\|'` is a non-raw string: `\|` is not a valid escape, so it stays as `\` + `|`, and `csv.Sniffer` receives candidates `, ; \t \ |` — a literal backslash becomes detectable as a delimiter, and Python 3.12+ emits `SyntaxWarning: invalid escape sequence '\|'` at import.

**Fix:**
```python
dialect = sniffer.sniff(sample, delimiters=r',;\t|')
```
(`|` needs no escape inside a `delimiters` string.)

This is also the one new function that is pure-Python with no QGIS dependency — a 4-case unit test (`,` `;` tab `|`, plus the `csv.Error → ','` fallback) would be the project's first runnable test and would have caught this.

File: plugin/mgp_config_reader.py [hunk 5]
```diff
@@ -71,6 +71,16 @@ def readConfig(self) -> typing.NoReturn:
                 if 'fileReferenceLayerField' in configReader['ReferenceLayer']:
                     self.mainWindow.ui.referenceLayerFieldCombobox.setCurrentIndex(int(configReader['ReferenceLayer']['fileReferenceLayerField']))
 
+            if 'UrbanismLayer' in configReader:
```
Comment (jpazvd) [inline plugin/mgp_config_reader.py:74]: **🟠 An "unofficial" feature is reachable from a config file while the UI is hidden.**

`UrbanismValidator.py` is headed "not part of any release … should not be considered as any official feature," and the Step-2 checkbox is hidden via `setVisible(False)`. But the validator is constructed in `CentroidsDisplacer.__init__` and `validateDisplacement()` is an active gate in `__displaceCentroid` (the loop exits only on `ref_id_after == ref_id_before and urbanism_valid`). This block re-checks the hidden checkbox and sets the raster path from a config file, so an `.ini` with `urbanismEnabled=True` + a raster silently activates an un-reviewed constraint that alters anonymisation — and when `MAX_ITERATIONS` is hit the loop emits the last candidate point (potentially constraint-violating) with only a text remark.

**Fix:** while the feature is hidden, don't read the `UrbanismLayer` section here (or split the whole DEGURBA path into its own branch/PR); define the iteration-exhausted behaviour so a non-compliant point is never emitted silently; and remove the stray `print("[UrbanismValidator] Displacement validation FAILED")` in `UrbanismValidator.validateDisplacement`. Either way, declare in the PR body that this code is shipped-but-hidden.

**Review by jpazvd — CHANGES_REQUESTED** [review]
## Multi-agent review — requesting changes

This PR was reviewed with a structured panel (code-quality, conventions, test-plan, and an adversarial pass), synthesized below. The work is fundamentally sound — the CSV/covariate refactors and the `QVariant` → `QMetaType` modernization are good — but there are **two confirmed runtime crashes in the anonymisation core** that should be fixed before merge. The path to green is short; the five inline comments mark the exact fixes.

### 🔴 Blockers (confirmed crashes)

**1. Declared `qgisMinimumVersion=3.0` is wrong — guaranteed crash below QGIS 3.38.**
The `QgsField(name, QMetaType.Type.X)` constructor was added in **QGIS 3.38** (the `QVariant` form was only *deprecated* there), and `Utils.writeLayerIfExists` now calls `writeAsVectorFormatV3` (**QGIS 3.20+**). On any QGIS 3.0–3.37 the plugin raises `AttributeError` at the first layer creation — i.e. the first *Generate Centroids* / *Displace Centroids* click. The wrong floor is also inside the hand-edited `metadata.txt` in `releases/micsgeocodeplugin-1.4.0.zip`, so an older install passes the version gate and then dies at step 1. (Inline 1.)

**2. Positional-argument bug in `getPoleOfInaccessibilityOrCentroid` crashes SHP loading on degenerate polygons.**
`CentroidsLoader.__loadCentroidsFromSHP` calls `getPoleOfInaccessibilityOrCentroid(cluster_polygon, self.pole_inaccessibility_precision)`, but the helper signature is `(polygon_feature, fallback_feature=None, precision=0.00001)` — so the precision float binds to `fallback_feature`. On the null-pole fallback branch this runs `0.00001.geometry()` → `AttributeError`, exactly for the spiky/degenerate polygons the pole method is meant to handle. The sibling call site (multipoint path) already uses keywords correctly. (Inline 2.)

### 🟠 Should-fix

- **Wrong row number + inconsistent error handling** in `CovariatesProcesser.computeCovariates`: the `raise ValueError("In row {rowIndex+1}…")` uses the stale `rowIndex` (left at the last row) instead of the per-row `row_index`; a count mismatch hard-aborts the whole batch while unknown-format / unsupported-stat soft-skip with a warning; and the `variety` removal pops from `column_names` *before* the length check, which can `IndexError` on the very mismatch the check is meant to catch. (Inline 3.)
- **`detect_csv_delimiter` candidate string** `delimiters=',;\t\|'` — `\|` is an invalid escape, so `\` becomes a candidate delimiter (and Python 3.12+ emits a `SyntaxWarning`). Use a raw string. (Inline 4.)
- **DEGURBA / `UrbanismValidator` is shipped hidden but live-wired** into the displacement loop and **re-activatable from a config file** (`mgp_config_reader` reads `UrbanismLayer.urbanismEnabled`), despite a header declaring it "not part of any release." It can also emit a non-compliant point silently when `MAX_ITERATIONS` is exhausted, and leaves a stray `print()`. Gate it off cleanly or split it out — and declare it in the PR body. (Inline 5.)

### 🟡 Also worth addressing

- **Methodology change not flagged:** pole-of-inaccessibility replaces the centroid as the displacement / nearest-distance reference, so 1.4.0 outputs are not reproducible against 1.3.0. Add a short equivalence note to the changelog / `USER_GUIDE.md`, and ideally a regression check on a known dataset.
- **Progress bar won't repaint:** the loops run synchronously on the GUI thread with no `QApplication.processEvents()`, so the bar jumps hidden → 100%. `value=24` also ships as the design-time default. Consider a `QgsTask` or a `processEvents()` after each update, and set the initial value to 0.
- **No test plan / no real tests:** `plugin/test/` is Plugin-Builder boilerplate that imports a nonexistent `MGP_dialog` and would error on collection. `detect_csv_delimiter` is pure-Python and a natural first unit test. A `## Test plan` enumerating the manual-in-QGIS steps would help reviewers verify the three bugfixes.
- Stray `# error here` comment in `CovariatesProcesser`; regenerate the release zip from `plugin/` via `WIN_zip.bat` rather than hand-editing it.

### ✅ Done well

Loop restructure in `__displaceCentroid` is behaviour-preserving (verified by tracing both versions); `__generateRemark` now accumulates all applicable remarks instead of overwriting only the last; `zonal_stat` selects output columns by name (order-safe vs QGIS's column order); pole logic centralized into one `Utils` helper; the six 1.4.0 version strings are consistent; `release-notes.md` follows the existing Features/Bugfixes format.

Fix the two blockers and the should-fix cluster and this is an approve.

---
<sub>Review assisted by a multi-agent panel run via Claude Code — code-quality, conventions, test-plan, and adversarial reviewers, synthesized into this verdict.</sub>
